src/conduits/RwaCounduits3.integration.t.sol

```diff
@@ -0,0 +1,377 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+import "forge-std/Test.sol";
+import "ds-token/token.sol";
+import "ds-math/math.sol";
+import "ds-value/value.sol";
+import "dss-interfaces/Interfaces.sol";
+
+import {DssPsm} from "dss-psm/psm.sol";
+
+import {RwaInputConduit3} from "./RwaInputConduit3.sol";
+import {RwaOutputConduit3} from "../conduits/RwaOutputConduit3.sol";
+
+abstract contract RwaConduits3TestAbstract is Test, DSMath {
+    // Define both in constructor of derived contract
+    bytes32 ILK;
+    address psm;
+
+    address me = address(this);
+    address testUrn = vm.addr(420);
+
+    ChainlogAbstract changelog = ChainlogAbstract(0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F);
+    VatAbstract vat = VatAbstract(changelog.getAddress("MCD_VAT"));
+
+    DaiAbstract dai;
+    GemAbstract gem;
+
+    uint256 GEM_DECIMALS;
+    uint256 GEM_DAI_DIFF_DECIMALS;
+    uint256 ART_LEFT;
+    uint256 URN_INK;
+
+    uint256 PSM_TIN;
+    uint256 PSM_TOUT;
+
+    RwaInputConduit3 inputConduit;
+    RwaOutputConduit3 outputConduit;
+
+    function ray(uint256 wad) internal pure returns (uint256) {
+        return wad * 10**9;
+    }
+
+    function rad(uint256 wad) internal pure returns (uint256) {
+        return wad * 10**27;
+    }
+
+    function wad(uint256 rad_) internal pure returns (uint256) {
+        return rad_ / 10**27;
+    }
+
+    function getDaiInAmount(uint256 gemAmt) internal view returns (uint256) {
+        uint256 gemAmt18 = mul(gemAmt, GEM_DAI_DIFF_DECIMALS);
+        uint256 fee = mul(gemAmt18, PSM_TOUT) / WAD;
+        return add(gemAmt18, fee);
+    }
+
+    function gemToDai(uint256 gemAmt) internal view returns (uint256) {
+        uint256 gemAmt18 = mul(gemAmt, GEM_DAI_DIFF_DECIMALS);
+        uint256 fee = mul(gemAmt18, PSM_TIN) / WAD;
+        return sub(gemAmt18, fee);
+    }
+
+    function setUp() public virtual {
+        dai = DaiAbstract(DssPsm(psm).dai());
+        gem = GemAbstract(AuthGemJoinAbstract(address(DssPsm(psm).gemJoin())).gem());
+
+        (uint256 art, , , uint256 line, ) = vat.ilks(ILK);
+        GEM_DECIMALS = 10**uint256(gem.decimals());
+        GEM_DAI_DIFF_DECIMALS = 10**uint256(dai.decimals() - gem.decimals());
+        ART_LEFT = (wad(line) - art) / GEM_DAI_DIFF_DECIMALS;
```

Comment:
This is a semantic error, if we are talking about `art` it should be in 18 precision and not gem one. I'd say we do the division when using it or rename to something which is proper.

Reply:
Done